common/interpreter/gipsy_scheme/packages/catch-package.scm

```diff
@@ -27,22 +27,37 @@
      (for-each (lambda (a) (write a)) args)
      (newline))
 
-   (define (throw . x)
+   (define (throw-with-no-continuations . x)
+     (if (_more-handlers?)
+         (begin (apply (_pop-handler) x) (quit -1))
+         ;;(apply (_pop-handler) x)
+         (apply error x)))
+
+   (define (throw-with-continuations . x)
      (if (_more-handlers?)
          (apply (_pop-handler) x)
          (apply error x)))
 
-   (macro (catch form)
+   (macro (catch-with-no-continuations form)
+     `(begin (catch-throw::_push-handler ,(cadr form)) ,@(cddr form)))
```

Comment:
you are currently not popping the handler in the good case which would seem to prevent proper catch-rethrow. Applying the same "(let ((,label ...." as done in catch-with-continuations" should enable this pattern?

Reply:
done